### DISCUSSION TITLE
data.lifecycle understanding

### DISCUSSION THREAD

**Discussion by holodnii** [gilbarbara/react-joyride#1016] [Q&A] [answered]
I'm trying to make Joyride work perfectly with my app, however facing some issues while integrating it in a controlled mode. The step logics works good, however all isues i have - are regarding fine step rendering. Elaborating: when i proceed from the step which has to trigger/dispatch some app component before showing the next step, I face the following: previous step does not get _"unmounted"_ before the app-component is dispatched. I tried working around `data.lifecycle`, which i thought seemed as a problem to this, but the sandbox code for the controlled mode does not even manage it's state (**in handleJoyrideCallback and onClick for the menu button**), [sandbox-controlled-mode](https://codesandbox.io/p/devbox/github/gilbarbara/react-joyride-demo?file=%2Fsrc%2FControlled%2Findex.tsx%3A167%2C17) so rn I'm kind of confused on where exactly should I start?
    
my onClick somewhere for something in app:
  
```javascript
 const onDetailsClicked = (): void => {
    if (isTouring) {
      if (currentStepIndex === 1) {
        // make perfect closing for step 2
        dispatch(setJoyrideLifecycle("ready")); // dispatch(setJoyrideLifecycle("complete"));?
        setTimeout(() => {
          dispatch(openAlertDialogAction(alert));
        }, 400);
        setTimeout(() => {
          dispatch(setCurrentStepIndex(currentStepIndex + 1));
        }, 1000); // ideal between 600-1000
        // last it should be
      } else if (currentStepIndex === 4) {
        setTimeout(() => {
          dispatch(openAlertDialogAction(alert));
        }, 400);
        setTimeout(() => {
          dispatch(setCurrentStepIndex(currentStepIndex));
        }, 1000); // ideal between 600-1000
      }
    }
```

**Comment by gilbarbara** [comment, marked as answer]
Hey @holodnii 

I'm not sure I understand the problem. Why are you setting the lifecycle?
Anyway, you need to listen for the lifecycle changes in the callback, stop the tour, and when your screen is ready, update the stepIndex and start the tour again.

**Reply by holodnii** [reply]
@gilbarbara Thank you for the rapid answer! I thought you need to as well control the lifecycle in controlled mode, but after reading the docs one more time, I found that Joyride props take only index for the controlled mode.
